crates/pop-telemetry/src/lib.rs

```diff
@@ -0,0 +1,365 @@
+use std::{
+	fs::{create_dir_all, File},
+	io,
+	io::{Read, Write},
+	path::{Path, PathBuf},
+};
+use thiserror::Error;
+
+const WEBSITE_ID: &str = "3da3a7d3-0d51-4f23-a4e0-5e3f7f9442c8";
+const CARGO_PKG_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[derive(Error, Debug)]
+pub enum TelemetryError {
+	#[error("a reqwest error occurred: {0}")]
+	NetworkError(reqwest::Error),
+	#[error("io error occurred: {0}")]
+	IO(io::Error),
+	#[error("opt-in is not set, can not report metrics")]
+	NotOptedIn,
+	#[error("telemetry is not set and no message will be sent")]
+	TelemetryNotSet,
+	#[error("unable to find config file")]
+	ConfigFileNotFound,
+	#[error("serialization failed: {0}")]
+	SerializeFailed(String),
+}
+
+pub type Result<T> = std::result::Result<T, TelemetryError>;
+
+#[derive(Debug, Clone)]
+pub struct Telemetry {
+	// Endpoint to the telemetry API.
+	// This should include the domain and api path (e.g. localhost:3000/api/send)
+	endpoint: String,
+	// Has the user opted-in to anonymous telemetry
+	opt_in: bool,
+	// Reqwest client
+	client: Client,
+}
+
+impl Telemetry {
+	pub fn new(endpoint: String, config_path: PathBuf) -> Self {
+		let opt_in = Self::check_opt_in_from_config(&config_path);
+
+		Telemetry { endpoint, opt_in, client: Client::new() }
+	}
+
+	fn check_opt_in_from_config(config_file_path: &PathBuf) -> bool {
+		let config: Config = match read_json_file(config_file_path) {
+			Ok(config) => config,
+			Err(err) => {
+				log::debug!("{:?}", err.to_string());
+				return false;
+			},
+		};
+
+		config.opt_in.allow
+	}
+
+	/// Send JSON payload to saved api endpoint.
+	/// Will return error and not send anything if opt-in is false.
+	/// Will return error from reqwest if the sending fails.
+	/// It sends message only once as "best effort". There is no retry on error
+	/// in order to keep overhead to a minimal.
+	async fn send_json(&self, payload: Value) -> Result<()> {
+		if !self.opt_in {
+			return Err(TelemetryError::NotOptedIn);
+		}
+
+		let request_builder = self.client.post(&self.endpoint);
+
+		request_builder
+			.json(&payload)
+			.send()
+			.await
+			.map_err(TelemetryError::NetworkError)?;
+
+		Ok(())
+	}
+}
+
+/// Generically reports that the CLI was used to the telemetry endpoint.
+/// There is explicitly no reqwest retries on failure to ensure overhead
+/// stays to a minimum.
+pub async fn record_cli_used(maybe_tel: Option<Telemetry>) -> Result<()> {
+	let tel = maybe_tel.ok_or(TelemetryError::TelemetryNotSet)?;
+
+	let payload = generate_payload("", json!({}));
+
+	let res = tel.send_json(payload).await;
+	log::debug!("send_cli_used result: {:?}", res);
+
+	res
+}
+
+/// Reports what CLI command was called to telemetry.
+///
+/// parameters:
+/// `command_name`: the name of the command entered (new, up, build, etc)
+/// `data`: the JSON representation of subcommands. This should never include any user inputted
+/// data like a file name.
+pub async fn record_cli_command(
+	maybe_tel: Option<Telemetry>,
```

Comment:
As above, remove the option, and probably the TelemetryNotSet error as expect it will become unused.